jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientStreamTest.java

```diff
@@ -1193,6 +1197,98 @@ public boolean handle(Request request, org.eclipse.jetty.server.Response respons
         assertTrue(clientLatch.await(timeoutInSeconds, TimeUnit.SECONDS));
     }
 
+    @ParameterizedTest
+    @MethodSource("transports")
+    public void testUploadWithRetainedData(Transport transport) throws Exception
+    {
+        List<Content.Chunk> chunks = new ArrayList<>();
+
+        start(transport, new Handler.Abstract()
+        {
+            @Override
+            public boolean handle(Request request, org.eclipse.jetty.server.Response response, Callback callback)
+            {
+                Runnable retainer = new Runnable()
```

Comment:
You may find `o.e.j.util.CompletableTask` better to use instead of bare `Runnable`.